doc/js-perspective-on-c.md

```diff
@@ -0,0 +1,108 @@
+
+```c++
+SomethingFancy::SomethingFancy(SomethingFancy &&other);
+SomethingFancy &SomethingFancy::operator=(SomethingFancy &&other);
+```
+
+Just knowing that those exist and that they are NOT compiler generated is two pieces of information that do not fit into either the `C` or `JS` mental model.
+
+The following lines are NOT The same...
+
+```c++
+SomethingFancy a;
+SomethingFancy b{42};
+SomethingFancy c = SomethingFancy{123};
+```
+
+SomethingFancy 'a' is allocated and constructed using the "default constructor" and in this instance that is a "magic function" that is compiler created. 'b' is allocated and created in-place using the declared constructor as would be expected. 'c' is... complex. The compiler-created default constructor is being used to allocate memory at 'c' and a right-hand reference is created using the declared constructor.  Then the compiler-created destructor deletes the temporary at 'c' leaving an empty allocation, and then the value in the right-hand reference is copied to the allocation at 'c' using the compiler-created "magic" copy assignment operator.  Lots and lots of confusing compiler slight-of-hand.
+
+While this is not intended to be a full tutorial these idea were something that definitely took getting used to and those "magic" functions are very hard to debug.  I would also like to make note that the above example is strictly a `c++` thing.  `SomethingFancy` is not a built in type and constructor syntax with an initializer list are not `c`, they are '`c++`'.  The ideas of right-hand reference, operator overloading and constructor/destructors are all "`++`" too.
+
+```c
+int a;
+int b{42}; // not valid c, but is valid c++
+int c = 42;
+```
+
+It is valid, and often times required, to allocate memory but to no initialize it.  In JS the uninitialized value is `undefined` in `c` using an uninitialized variable is "undefined behavior".  
+
+With `c` as the context 'a' is uninitialized and can be dangerous. 'b' is invalid syntax and 'c' is an integer initialized to 42.
+
+With `c++` as the context 'a' is treated the same as raw `c`. 'b' is an integer initialized to 42 using constructor syntax. 'c' is an integer initialized to 42 using assignment syntax.
+
+## Passing To and Returning From Functions
+
+Building functions happen for the same reasons in c as they do in JS.  They help to ease the mental burden for "what something does".  The differences in memory allocation drive very different coding paradigms though.
+
+In modern JS we generally pass in what is needed inside the function context, return back the values that result and if an error occurs it is thrown.  That pattern generally holds true for both sync an async functions.
```

Comment:
```suggestion
In modern JS we generally pass in what is needed inside the function context, return back the values that result, and if an error occurs, it is thrown.  That pattern generally holds true for both sync and async functions.
```

Reply:
Updated.  I played with adding and removing the second `,` and went with not including so that phrase flows.  What do you think?